### ISSUE TITLE
Getting Playback Error: This server is not powerful enough to convert video.

### ISSUE THREAD

**Issue by MagicLegend** [UnicornTranscoder/UnicornTranscoder#89]
Hi,

I have done my best to configure my plex (running on a pi 3b+) to offload the transcoding tasks (which it obviously sucks at) to my other more powerful machine(s). I have created the following layout:

Pi: Plex server, UnicornLoadBalancer, Nginx reverse proxy to handle TLS and stuffs
Other machine: Media library (mounted to the pi), pure Debian VM for the UnicornTranscoder with the `3000` port forwarded into it from the host (so should make no difference)

The transcoder works, and is accessible via both it's internal ip (`localhost:3000` from the **host**, to verify the portforward would work) and it's external domain. Gives the neat 'You did things right after spending the night setting this up` purple page.

I have dropped port 32400 on the pi, moved the `/usr/lib/plexmediaserver/Plex Relay` file and replaced the `/usr/lib/plexmediaserver/Plex Transcoder` with the generated one.

However. When I try to start a transcode task (aka running a high-bitrate movie from a browser), I get the error:

```
Playback Error:
This server is not powerful enough to convert video.
```

Ehh what now? I am not really sure what logs to provide you with (nor really how to get them out of the software :-)), but any guidance would be much appreciated! The error would make me guess that I did in fact change the `Plex Transcoder` to something different; but it's getting confused at recognizing the running transcoder? I can at least confirm that the transcoder is receiving no requests by keeping an eye on the nginx logs.

The next step would be for it to try to probe multiple machines for the transcoding tasks (e.g. multiple desktops/laptops around the house to see which ones are up); is this something unicorn can handle?

**Comment by MagicLegend** [comment]
I just read in another issue that the transcoders also need access to the media files (pretty obvious now I think of it), which the transcoder does not have right now. However; since there is no request going to the transcoder domain at all, I would suggest this is not part of the issue right now.

**Comment by Maxou44** [comment]
It looks like a plex issue, can you share us your plex transcode settings ? Did you disabled or limit transcodes on Plex Media Server ?

**Comment by Maxou44** [comment]
The transcoder need to talk with the load balancer address, by default the load balancer will send files if needed

**Comment by MagicLegend** [comment]
Hi Max,

Thanks for the quick response! I really wanna get this working :)

The settings: 
![image](https://user-images.githubusercontent.com/3169104/79687329-dae53f00-8246-11ea-9d26-a925232f4232.png)

I can see in the access logs that from my own external ip there are requests to `/api/sessions` on the transcoder. After a first `200` return code; this only returns `304`; as I would expect. This, combined with that the transcoder is accessable externally; would tell me that the transcoder is working.

It should be noted that http requests aren't properly upgraded for some reason. Using the configs found in your repo; the server block that should handle the `301` actually doesn't. Nginx returns an error that a plain HTTP request was sent to an HTTPS port. Not sure why.


> The transcoder need to talk with the load balancer address, by default the load balancer will send files if needed
So you are saying that the loadbalancer (the pi in this case) would only need access to the share? Not the transcoder itself?

**Comment by Maxou44** [comment]
Can you switch "Maximum simultaneous transcode" to the maximum/unlimited value ?
From the Plex web, can you start a transcode and show me the network requests ?
Can you share us transcoder and load balancer configurations ?

**Comment by MagicLegend** [comment]
Actually, upping the `maximum simultaneous transcode` removes the warning that is given. Now it just gets stuk in a waiting loop.

Transcoder config:
```
/**
 * Created by drouar_b on 27/04/2017.
 */

const env = require('getenv');

module.exports = {
    port:                  env.int   ('SERVER_PORT',           3000),
    host:                  env.string('SERVER_LISTEN',         '127.0.0.1'),
    transcoder_decay_time: env.int   ('TRANSCODER_DECAY_TIME', 120),
    loadbalancer_address:  env.string('LOADBALANCER_ADDRESS',  'https://plex.redacted.com'),
    ping_frequency:        env.int   ('PING_FREQUENCY',        10),
    instance_address:      env.string('INSTANCE_ADDRESS',      'https://transcoder.plex.redacted.com'),

    maxmind_key:           env.string('MAXMIND_KEY',           ''),

    transcoder: {
        plex_arch:         env.string('PLEX_ARCH',             'amd64'), // Can be: amd64, arm64, armhf, i386
        plex_build:        env.string('PLEX_BUILD',            '1.19.1.2645-ccb6eb67e'),
        codecs_build:      env.string('CODECS_BUILD',          '99c90e0-3095'),
        eae_version:       env.string('EAE_VERSION',           'eae-69c1de6-42'), // 41 or 42?
        plex_resources:    env.string('PLEX_RESOURCES',        'plexmediaserver/'),
        temp_folder:       env.string('TEMP_FOLDER',           'cache/'),
        codecs_folder:     env.string('CODECS_FOLDER',         'codecs/'),
        plex_transcoder:   env.string('PLEX_TRANSCODER',       'Plex Transcoder'),
        debug:             env.boolish('TRANSCODER_DEBUG',     false),
        debugScope:        env.boolish('DEBUG',                false),
    },

    performance: {
        maxSessions:        env.int('MAX_SESSIONS',           10),
        maxDownloads:       env.int('MAX_DOWNLOADS',          10),
        maxTranscodes:      env.int('MAX_TRANSCODE',          10),
    },

    //routing: {
    //    'US': 'http://usgateway.myplex.com'
    //},
};
```

I have not in fact verified all the version numbers in here; only that the plex version matched. I figured I would run into issues with transcoding if these were off. (but not that the request to transcode fails)

What happens when I start a movie (from a timestamp):

![image](https://user-images.githubusercontent.com/3169104/79695764-f6693d80-8278-11ea-809c-2b5a06d991ae.png)

Starting this generated 0 traffic on the transcoder domain, as seen in the access logs of the Nginx proxy. The access logs of the plex server domain from nginx (and thus also of the load balancer if I'm not mistaken):

```
1.2.3.4 - - [19/Apr/2020:20:01:38 +0200] "GET / HTTP/2.0" 401 157 "-" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:38 +0200] "GET / HTTP/2.0" 401 157 "-" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:38 +0200] "GET /web/index.html HTTP/2.0" 200 3576 "https://plex.redacted.com/" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:38 +0200] "GET /web/chunk-2-9386d0a8ac3648330186-plex-4.22.3-79842ff.css HTTP/2.0" 200 977550 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:38 +0200] "GET /web/js/chunk-2-9386d0a8ac3648330186-plex-4.22.3-79842ff.js HTTP/2.0" 200 4217725 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /web/translations/en-GB.json HTTP/2.0" 200 11788 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /web/common/img/backgrounds/preset-dark2.6d2c8336496f761f38de8558a2b9709b.png HTTP/2.0" 200 113817 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /web/common/img/backgrounds/noise.8b05ce45d0df59343e206bc9ae78d85d.png HTTP/2.0" 200 54413 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /media/providers?X-Plex-Product=Plex%20Web&X-Plex-Version=4.22.3&X-Plex-Client-Identifier=bj0f2p5ukopjnnnxooxnufa2&X-Plex-Platform=Chrome&X-Plex-Platform-Version=80.0&X-Plex-Sync-Version=2&X-Plex-Features=external-media%2Cindirect-media&X-Plex-Model=bundled&X-Plex-Device=Windows&X-Plex-Device-Name=Chrome&X-Plex-Device-Screen-Resolution=1920x1057%2C1920x1200&X-Plex-Language=en-GB HTTP/2.0" 401 82 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /?X-Plex-Product=Plex%20Web&X-Plex-Version=4.22.3&X-Plex-Client-Identifier=bj0f2p5ukopjnnnxooxnufa2&X-Plex-Platform=Chrome&X-Plex-Platform-Version=80.0&X-Plex-Sync-Version=2&X-Plex-Features=external-media%2Cindirect-media&X-Plex-Model=bundled&X-Plex-Device=Windows&X-Plex-Device-Name=Chrome&X-Plex-Device-Screen-Resolution=1920x1057%2C1920x1200&X-Plex-Language=en-GB HTTP/2.0" 401 157 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /identity?X-Plex-Product=Plex%20Web&X-Plex-Version=4.22.3&X-Plex-Client-Identifier=bj0f2p5ukopjnnnxooxnufa2&X-Plex-Platform=Chrome&X-Plex-Platform-Version=80.0&X-Plex-Sync-Version=2&X-Plex-Features=external-media%2Cindirect-media&X-Plex-Model=bundled&X-Plex-Device=Windows&X-Plex-Device-Name=Chrome&X-Plex-Device-Screen-Resolution=1920x1057%2C1920x1200&X-Plex-Language=en-GB HTTP/2.0" 200 169 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:39 +0200] "GET /media/providers?X-Plex-Product=Plex%20Web&X-Plex-Version=4.22.3&X-Plex-Client-Identifier=bj0f2p5ukopjnnnxooxnufa2&X-Plex-Platform=Chrome&X-Plex-Platform-Version=80.0&X-Plex-Sync-Version=2&X-Plex-Features=external-media%2Cindirect-media&X-Plex-Model=bundled&X-Plex-Device=Windows&X-Plex-Device-Name=Chrome&X-Plex-Device-Screen-Resolution=1920x1057%2C1920x1200&X-Plex-Token=oWMs8yqYBhbCynV2F2Nc&X-Plex-Language=en-GB HTTP/2.0" 200 2597 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:41 +0200] "GET /web/common/img/backgrounds/preset-light2.ead08af5122abea1ab8f7d601fd93d6a.png HTTP/2.0" 200 125754 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:01:43 +0200] "GET /web/chunk-2-9386d0a8ac3648330186-plex-4.22.3-79842ff.css HTTP/2.0" 200 977550 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:02:00 +0200] "GET /web/js/chunk-8-6adc7a59834e820b9990-plex-4.22.3-79842ff.js HTTP/2.0" 200 278582 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
1.2.3.4 - - [19/Apr/2020:20:02:00 +0200] "GET /web/js/chunk-6-1b61f28a8c710b19a8b1-plex-4.22.3-79842ff.js HTTP/2.0" 200 7168 "https://plex.redacted.com/web/index.html" "UserAgentStringThatIsNotImportantButIUsedChrome80"
```

The used nginx config:

```
server {
    # Listen
    listen 80;
    listen [::]:80;
    server_name plex.redacted.com;
    # Redirect
    location / {
        return 301 https://plex.redacted.com;
    }
}
server {
    # Listen
    listen 443 ssl http2;
    listen [::]:443 ssl http2;
    server_name plex.redacted.com;

    # Logs
    error_log /var/log/nginx/error.log;
    access_log /var/log/nginx/access.log;

    # SSL
    ssl_certificate /redacted.com.crt;
    ssl_certificate_key /redacted.com.key;

    # Timeout
    client_body_timeout 240m;
    send_timeout 100m;
    # Proxy
    location / {
        # Proxy configuration
        proxy_pass http://127.0.0.1:3001;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

        # Proxy settings
        proxy_redirect      off;
        proxy_buffering     off;
        ## Websockets support
        proxy_http_version      1.1;
        proxy_set_header        Upgrade         $http_upgrade;
        proxy_set_header        Connection      "upgrade";
        proxy_read_timeout      36000s;
        proxy_send_timeout      36000s;
        proxy_connect_timeout   36000s;
    }
}

server {
    listen 80;
    listen [::]:80;
    server_name transcoder.plex.redacted.com;

    location / {
            return 301 https://transcoder.plex.redacted.com;
    }
}

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;
    server_name transcoder.plex.redacted.com;
    error_log /var/log/nginx/transcoder-error.log;
    access_log /var/log/nginx/transcoder-access.log;
    ssl_certificate /_.plex.redacted.com.crt;
    ssl_certificate_key /_.plex.redacted.com.key;
    client_body_timeout 240m;
    send_timeout 100m;

    location / {
        allow all;
        proxy_set_header        Host $host;
        proxy_set_header        X-Real-IP $remote_addr;
        proxy_set_header        X-Forwarded_For $proxy_add_x_forwarded_for;
        proxy_pass              http://192.168.1.1:3000/;
        proxy_buffering         off;
        proxy_read_timeout      36000s;
        proxy_send_timeout      36000s;
        proxy_connect_timeout   36000s;
    }
}
```

The loadbalancer config is not changed; since the defaults all seemed sensible to me. (Also, why aren't these just placed in a `config.js` like the rest? :))

**Comment by drouarb** [comment]
Hey,
I see that you got only one start.mpd, it looks like you are not reaching your Plex through the UnicornLoadBalancer.
In your dev tool you should first see a request on your Plex domain with a 302 answer redirecting you to the transcoder.

**Comment by MagicLegend** [comment]
Hey,

I just restarted all the services, and I'm not getting the error anymore! That's a win I guess. However, now it just keeps spinning and spinning until it eventually times out. I pinned it down to the "Maximum Simultaneous Transcode" option only being set to 1.

However I am certain the request goes through the load balancer: 
![image](https://user-images.githubusercontent.com/3169104/79998229-b8f3f280-84ba-11ea-9cb1-939573486416.png)

This is me starting a stream. However, I can see that no request actually makes it through to the transcoder (or at least not via the public domain); since nothing shows up in the nginx logs for the transcoder. And from the loadbalancer I can't see if the request is external or not. Could it be that the transcoder does not identify itself to the loadbalancer? I shouldn't have to hardcode the transcoder address in the loadbalancer config right?

**Comment by drouarb** [comment]
Yes, the transcoder should not be configured in the LoadBalancer, normally, in your transcoder configuration, if LOADBALANCER_ADDRESS is set, the transcoder will register itself to the LoadBalancer. Then, it will tell the LoadBalancer that it can be pinged on the INSTANCE_ADDRESS

**Comment by MagicLegend** [comment]
Using what endpoint should the transcoder register itself? Then I can monitor the logs / network traffic for that call. How can I get debug info out of the transcoder? Giving `DEBUG=UnicornTranscoder*` as per the docs results in an exception that it wants a boolean.

```
magiclegend@debian:~/unicorn/UnicornTranscoder$ DEBUG=UnicornTranscoder npm start

> UnicornTranscoder@2.2.2 start /home/magiclegend/unicorn/UnicornTranscoder
> node ./bin/www

/home/magiclegend/unicorn/UnicornTranscoder/node_modules/getenv/index.js:70
        throw new Error('GetEnv.NoBoolean: ' + value + ' is not a boolean.');
        ^

Error: GetEnv.NoBoolean: UnicornTranscoder is not a boolean.
    at Object.boolish (/home/magiclegend/unicorn/UnicornTranscoder/node_modules/getenv/index.js:70:15)
    at Function.boolish (/home/magiclegend/unicorn/UnicornTranscoder/node_modules/getenv/index.js:84:27)
    at Object.<anonymous> (/home/magiclegend/unicorn/UnicornTranscoder/config.js:27:32)
    at Module._compile (internal/modules/cjs/loader.js:1156:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1176:10)
    at Module.load (internal/modules/cjs/loader.js:1000:32)
    at Function.Module._load (internal/modules/cjs/loader.js:899:14)
    at Module.require (internal/modules/cjs/loader.js:1042:19)
    at require (internal/modules/cjs/helpers.js:77:18)
    at Object.<anonymous> (/home/magiclegend/unicorn/UnicornTranscoder/core/transcoder.js:10:16)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! UnicornTranscoder@2.2.2 start: `node ./bin/www`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the UnicornTranscoder@2.2.2 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by drouarb** [comment]
DEBUG=* => Enable logging
TRANSCODER_DEBUG=true => Add FFMPEG Log to the log

**Comment by MagicLegend** [comment]
@drouarb Are you sure I shouldn't be calling `npm` differently? Because setting `DEBUG=*` results in the same error as above. Setting `DEBUG=true` also doesn't seem to have any effect.

**Comment by MagicLegend** [comment]
Okay, I can confirm that the transcoder is calling `/api/update` on the load balancer. Is there a debug page on the load balancer where I can see what jobs its trying to offload to which transcoder?

**Comment by drouarb** [comment]
Ok, I just looked at the code, a mistake was made when the debug page was merged. 
I fixed it in the latest release, try to install the latest version and you should be able to use the DEBUG environment as explained in the README.
Also, in the latest release, a debug web page has been added, you can access it on the / of your transcoder.

**Comment by MagicLegend** [comment]
Hi,

No worries. Happens to the best of us! I pulled the transcoder, and now the `DEBUG=UnicornTranscoder* npm start` does what I expect. However, there is still no logging when I visit the webpage at `transcoder.plex.redacted.com`; is that intended? I do see the following output when I start the transcoder:

```
> UnicornTranscoder@2.5.1 start /home/magiclegend/unicorn/UnicornTranscoder
> node ./bin/www

  UnicornTranscoder Starting EAE +0ms
  UnicornTranscoder2:server Listening on port 3000 +0ms
```

However, still no logging or a change on the transcoder webpage when I try to start a movie that needs to be transcoded. The `start.mpd` request still goes to the `plex` server itself, and responds with a `200` after a few seconds; and not with the `302` I would expect based on your comments.

I would think that there is an bug in the loadbalancer _somewhere_; but I wouldn't really know where. How could I further debug this?

**Comment by MagicLegend** [comment]
I just noticed that in the `Dashboard`; the server thinks it's playing and transcoding:
![image](https://user-images.githubusercontent.com/3169104/80099761-485add80-856f-11ea-9ef2-b9716f32b380.png)

Note that this is the same machine that is also hosting the Transcoder VM. However, when I use my laptop on my phone hotspot (= forced externally); IT WORKS! I get a `307` with the transcoder, and I can see it showing up on the transcoder page.

So, I would conclude from this that something is up when you try to reach Plex from the internal network. I noticed that from within my network it uses an domain like `https://192-168-1-1.<HASH>.plex.direct:32400/video/:/transcode/universal/start.mpd`; whilst externally it uses the external address. I think this would be my misconfiguration? However, I have blocked port 32400 in the firewall of the Plex host; and I renamed the `Plex Relay` binary (it's still in the folder tho).

**Comment by MagicLegend** [comment]
Also I see the transcoder trying to reach `127.0.0.1:3001`; which I assume is it looking for the loadbalancer. However; the loadbalancer is not there (it's another machine). Is this because the loadbalancer thinks it's at `127.0.0.1:3001` because of the reverse proxy?

![image](https://user-images.githubusercontent.com/3169104/80101535-d20baa80-8571-11ea-9b2d-9990bc3427a5.png)

**Comment by drouarb** [comment]
If your request goes to plex.direct, it bypass the UnicornLoadBalancer, so it's normal that it does not work. All direct access to Plex Media Server should be denied. To do that, you can use iptables as metioned in
https://github.com/UnicornTranscoder/UnicornLoadBalancer#2-notes

For the 127.0.0.1:3001, it is because you are using CUSTOM_DOWNLOAD_FORWARD: false
In this configuration the UnicornLoadBalancer will serve files to the transcoder, so you don't have to mount your FS to the transcoder.
To use this option SERVER_PUBLIC should be set to a public address where the LoadBalancer will serve the files.

**Comment by MagicLegend** [comment]
Hmm, that's weird. That port is in fact already being dropped.

```
  % sudo iptables -L                                                                                                                                                                                                                                                                                                                                                  !637
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
DROP       tcp  --  anywhere             anywhere             tcp dpt:32400                                                                                                                                                                                                                                                                                                
Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```

Nmap reports the same:

```
$ nmap plex.redacted.com -p 32400
Starting Nmap 7.80 ( https://nmap.org ) at 2020-04-23 19:41 W. Europe Summer Time
Nmap scan report for plex.redacted.com (1.2.3.4)
Host is up (0.0070s latency).
rDNS record for 1.2.3.4: (*)

PORT      STATE    SERVICE
32400/tcp filtered plex

Nmap done: 1 IP address (1 host up) scanned in 12.31 seconds
```

However it still is able to connect somehow? I found the following commands when looking around for more solutions:

```
sudo chmod 0000 /usr/lib/plexmediaserver/Plex\ Relay
sudo chattr +ia /usr/lib/plexmediaserver/Plex\ Relay
```

This also doesn't seem to work. However, there is no relay binary started:

```
  % ps aux | grep plex                                                                                                                                                                                                                                                                                                                                                !679
plex     12750  1.7  6.5 297040 61652 ?        Ssl  19:58   0:11 /usr/lib/plexmediaserver/Plex Media Server
plex     12787  2.6  3.3 187132 32164 ?        SNl  19:58   0:17 Plex Plug-in [com.plexapp.system] /usr/lib/plexmediaserver/Resources/Plug-ins-ccb6eb67e/Framework.bundle/Contents/Resources/Versions/2/Python/bootstrap.py --server-version 1.19.1.2645-ccb6eb67e /usr/lib/plexmediaserver/Resources/Plug-ins-ccb6eb67e/System.bundle
plex     12835  0.0  1.2 107268 11400 ?        Sl   19:58   0:00 /usr/lib/plexmediaserver/Plex Tuner Service /usr/lib/plexmediaserver/Resources/Tuner/Private /usr/lib/plexmediaserver/Resources/Tuner/Shared 1.19.1.2645-ccb6eb67e 32600 /waitmutex
plex     12895  1.7  3.0 136584 29040 ?        Sl   19:58   0:11 Plex Plug-in [com.plexapp.agents.imdb] /usr/lib/plexmediaserver/Resources/Plug-ins-ccb6eb67e/Framework.bundle/Contents/Resources/Versions/2/Python/bootstrap.py --server-version 1.19.1.2645-ccb6eb67e /usr/lib/plexmediaserver/Resources/Plug-ins-ccb6eb67e/PlexMovie.bundle
plex     12945  0.0  0.1   5464  1312 ?        S    19:59   0:00 Plex EAE Service
```

How is it still doing this? I have also disabled the setting at settings > network > Enable Relay.

As for the other setting; I missed that, my bad :-)

**Comment by drouarb** [comment]
Your nmap says that the port is closed. I don't know how you bypassed the firewall.
To find out, you should look in the devtool the address and the domain used by your client.

**Comment by MagicLegend** [comment]
I think I figured it out. The pi has two network interfaces, the wifi and cable. After I set the `iptables` rule for both interfaces it started to 'work'. Externally the transcoder purrs like a kitten. However, when I'm internally, I see it giving me the `307` to the transcoder, the request to the transcoder gets made, but it never reaches the transcode server (nothing shows up in the logs, whereas the logs get spammed when an external stream gets started). I've given it a few minutes, but it never seems to do anything. It needs some more debugging this weekend :-)

Thank you all for your time! I'll keep this issue posted on my progress.

tl;dr: set the DROP rule for all your INCOMING traffic on ALL interfaces.

**Comment by MagicLegend** [comment]
So, next issue. Internally I see the following:

- start.mpd request to the plex server > responds with 307
- start.mpd request to the transcoder > shows up in the nginx logs

However, the second request actually doesn't do anything. Enabling all the debug logging on the transcoder shows express receiving a request: `query  : /api/resolve?session=asdfasdfdasfdsfadsf&ip=1.2.3.4&origin=https%3A%2F%2Fplex.redacted.com%2F +0ms` However, no stream gets started, and the second request never seems to get an answer according to the browser. As I said before, externally (= outside the network of both the plex server and the transcoder) it all works perfectly.

Is this indeed the endpoint that should be called? Are there more requests it should make?

**Comment by MagicLegend** [comment]
Also I noticed that my iOS app stopped recognizing my own server. It just permanently shows as offline, even though I can reach it just fine from a mobile browser. Is this known behavior?

**Comment by Maxou44** [comment]
> Also I noticed that my iOS app stopped recognizing my own server. It just permanently shows as offline, even though I can reach it just fine from a mobile browser. Is this known behavior?

You need to set your domain name in Plex settings 😉

**Comment by MagicLegend** [comment]
You mean `settings > network > custom server access urls`? Because that option was never set (I don’t want it to be published). I guess the app found the plex.direct workaround or smth, and that’s why it still worked...

**Comment by Maxou44** [comment]
The plex.direct can't work with the unicorn transcoder, if you have access to it using plex direct the 32400 port is not correctly blocked.

**Comment by MagicLegend** [comment]
Hi Max,

I am aware that `plex.direct` doesn't work with unicorn. Makes sense. As I said 2 days ago; I had made a mistake in not blocking both interfaces on the pi. The port blocking now works correctly. However, the Plex app stopped recognizing the server; it's shown as permanently offline now. I wish Plex stopped with the automatically connecting things; I want to do that myself...

There is a difference between `plex.direct` and my own domain. I have full control over my own domain. I have no control over what calls home the plex server makes. I'm not a fan of that. I want to further add security to the nginx proxy (client certs with whitelists for the mobile apps). However, that is beyond why this issue is still open.

Please have a look at my question again:

> So, next issue. Internally I see the following:
> 
> * start.mpd request to the plex server > responds with 307
> * start.mpd request to the transcoder > shows up in the nginx logs
> 
> However, the second request actually doesn't do anything. Enabling all the debug logging on the transcoder shows express receiving a request: `query : /api/resolve?session=asdfasdfdasfdsfadsf&ip=1.2.3.4&origin=https%3A%2F%2Fplex.redacted.com%2F +0ms` However, no stream gets started, and the second request never seems to get an answer according to the browser. As I said before, externally (= outside the network of both the plex server and the transcoder) it all works perfectly.
> 
> Is this indeed the endpoint that should be called? Are there more requests it should make?

**Comment by Maxou44** [comment]
For me your transcoder can't contact the Load Balancer

**Comment by MagicLegend** [comment]
I can confirm that that is not true. When I start the loadbalancer (= cleared it's cache), `/api/stats` shows no transcoder. When I start the transcoder it shows up.

**Comment by Maxou44** [comment]
The start.mpd redirects where ?

**Comment by MagicLegend** [comment]
**Working from an external IP**
![image](https://user-images.githubusercontent.com/3169104/80315323-0e037180-87f7-11ea-9624-3999eb69ab64.png)
This stream actually started and is getting transcoded.

**Broken from within the same network**
![image](https://user-images.githubusercontent.com/3169104/80315339-23789b80-87f7-11ea-820b-ac53ee27f566.png)
The first request is to the main plex server, which responds with a `307` to the transcoder. The request to the transcoder gets made, but never gets a response.

---

On another note, I just found out that you can't log in anymore to the server... It tries to reach for `/media`, and after getting a `401` it just drops back to the login screen... I'm getting crazy from this software; I wish it would stop trying to be all automagically, and just gives me clear and precise errors... I'm already starting to consider to just give up on this :D

**Comment by MagicLegend** [comment]
I think I have figured out why the (desktop) apps don't work anymore; in the logs I notice it's trying to reach the `plex.direct` domain or the direct internal IP of the plex server... The drama never ends... When will this thing stop trying to think for me; and just prompt me to enter the right server domain?

**Comment by Maxou44** [comment]
For me the transcoder address is invalid, normally it should reply when you ask start.mpd from it

Apps will use custom domain it's set in plex settings

**Comment by MagicLegend** [comment]
> For me the transcoder address is invalid, normally it should reply when you ask start.mpd from it

For you where? Have you rebuilt my setup? Or in your own environment? I am pretty sure the address is valid; otherwise with these settings it wouldn’t have worked when I’m outside my own network right? That I am looping back to a VM on the same machine shouldn’t really matter from a networking-perspective.

The transcoder is accessible as follows:

- Request goes to nginx proxy (the same that also handles the plex server itself)
- Gets forwarded to the 192.168.x.y address of my desktop that hosts the VM (static ip)
- On my desktop the firewall allows port 3000 (or 3001; whichever it used :-))
- The VMWare NAT adapter will listen on said port and forward it into the VM
- The VM just simply has the transcoder started and listening

**Comment by Maxou44** [comment]
Maybe the load balancer can't call plex ?

**Comment by MagicLegend** [comment]
But if the load balancer couldn't call plex; plex shouldn't be accessible in any form right? But since it works externally I'm confused what could cause this... And the requests made from inside the network should be routed through the domain name, just like external requests. The only thing I can think of is that Plex is still recognizing that the server is on the same external IP, and that it's trying to reach the Plex server over the internal network?

I am strange :D It's really simple, this project was built for Debian/Ubuntu systems, and I still run Windows on my main computers (because of gaming and laziness; I work/develop in Linux VM's). To start with I wanted to get it working using the specs you guys provided (= Debian/Ubuntu based machine). I could migrate to WSL later down the line if I really wanted to, since it shares resources with the host (thus being able to access the full CPU for transcoding).

**Comment by Maxou44** [comment]
LB <=> Plex link doesn't have impact on Plex accessibility if you don't tweak it, Plex can use Plex Direct/Plex Relay to be available. I'm sure there is an issue in your setup, can you access to plex from the load balancer address?

- LB should talk with Plex
- Plex should talk with LB (and ONLY the LB, remove Plex Relay, use custom domain name, block 32400 port for ingoing connections). LB should also access to Transcoders
- Transcoders should talk with LB

**Comment by MagicLegend** [comment]
> I'm sure there is an issue in your setup

Yep, agreed :D

> can you access to plex from the load balancer address?

Yes, this is how I access the plex server. I can see this show up in the load balancer logs, and in the nginx logs. The nginx config is waay above, and I've just updated it to fix an little oopsie I made.

> remove Plex Relay

Confirmed to be done; relay binary is renamed and made not executable with `chmod` to be extra sure.

![image](https://user-images.githubusercontent.com/3169104/80373942-2bd3e380-8896-11ea-8384-560003b19473.png)

> use custom domain name

![image](https://user-images.githubusercontent.com/3169104/80373055-c6332780-8894-11ea-98dc-b40ece35a589.png)

> block 32400 port for ingoing connections)

![image](https://user-images.githubusercontent.com/3169104/80373233-0abec300-8895-11ea-9f04-59aa809fe8ad.png)

> LB should also access to Transcoders

As I've said before; this seems to be working just fine. The Transcoder shows up on the LB when I start it, which I would assume indicates a successful handshake between the two. Also I can confirm that the transcoder is accessible via the external domain (debug page shows up). Since this domain is identical to the one reported by the LB on `/api/stats`, I would assume this to work.

![image](https://user-images.githubusercontent.com/3169104/80373537-802a9380-8895-11ea-96c3-230e1724ec69.png)
_I just noticed that a transcoder that goes offline never gets removed from this list. Is that intentional? I would expect the LB to drop a transcoder after X amount of heartbeats fail / don't come through._

> LB should talk with Plex

Plex works externally, so I assume this to work.

> Plex should talk with LB

How would I go about testing this? As you can see above, the ports are blocked on the pi for outgoing traffic on both interfaces, so I would assume Plex can no longer talk over that port to the outside world.

> Transcoders should talk with LB

Again, transcoder shows up on the `/api/stats` page, so I assume this to work.

**Comment by Maxou44** [comment]
When you start a stream from Plex Web, the start.mpd is correctly redirected?
What is the result of the call on the transcoder? Can you open the call in a new target and see what happend?

**Comment by MagicLegend** [comment]
> When you start a stream from Plex Web, the start.mpd is correctly redirected?

Yes it is. At least to the right domain. I of course don't know if the endpoint / arguments are 100% correct.

> What is the result of the call on the transcoder?

Eventually a timeout from the proxy.

> Can you open the call in a new target and see what happend?

**If I open it in a new tab, it starts working... Yaay?** I can see the transcoder actually starting the transcoding process; but on the original tab nothing happens. Eventually the transcoder stops on it's own.

However, I also pinned this down to being only in Chrome (80.0) **In Firefox it all works.** However, in Firefox **incognito** I still have login issues (it just errors out without anything apparent in the logs). But that might also be because Plex can be a steaming pile of garbage sometimes :-) I've read about more people having this issue.

**Comment by Maxou44** [comment]
Maybe I have an idea, can you try to stream using incognito mode on Google Chrome and disable your browser plugins ?

**Comment by MagicLegend** [comment]
Damn, of course. Yeah, disabling all extensions ([using an extension](https://chrome.google.com/webstore/detail/disable-extensions-tempor/lcfdefmogcogicollfebhgjiiakbjdje?hl=en), as one should) works. Now the stream gets started properly... I shall invest some time later to find the culprit!

**Comment by MagicLegend** [comment]
After a little bit of testing I've found the culprit: Tampermonkey. I was afraid of that... Could you do me a favour and try to install the extension in a Chrome 80 and test if the `start.mpd` request to the **transcoder** gets through?

What's funny is that re-sending the request (DevTools > Network > Right click the second `start.mpd` request and `Open in new tab`) would still have all the extensions enabled. Why would that work, but not when Plex started it? By the way; opening the first `start.mpd` request (on `plex.redacted.com`; which gives the `307` in a new tab also works.

**Comment by MagicLegend** [comment]
Also I have noticed that my i7-7700K @ 4.9GHz has trouble transcoding fast enough. I am aware that the VM will add overhead, but it seems to me like the FFMPEG is just not being multi-threaded / efficient enough. The VM is on an SSD; so it's cache will be at least on that speed. Right now I managed to let the stream deplete the buffer within 4 minutes;  which doesn't sound right to me. I can see the VM (which has 6 cores out of 8) asking for about 33% of the CPU with peaks (it's not a constant load like I would expect).

For context, here is part of the logs:

```
  UnicornTranscoder:Transcoder  +0ms                                                                                                                                                                                                                                             UnicornTranscoder:Dash Serving chunk-stream1-00199.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream1-00200.m4s for session o5nrmi7g0aii6va25wr0d865 +270ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00200.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream0-00199.m4s for session o5nrmi7g0aii6va25wr0d865 +256ms                                                                                                                                                                             UnicornTranscoder:Transcoder FFMPEG(stderr):     Last message repeated 1 times                                                                                                                                                                                                 UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'sub-chunk-00015' for writing                                                                                                                                                                                        UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                                  UnicornTranscoder:Transcoder  +2s                                                                                                                                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream1-00200.m4s for session o5nrmi7g0aii6va25wr0d865 +570ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00200.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream0-00200.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream1-00200.m4s for session o5nrmi7g0aii6va25wr0d865 +1ms                                                                                                                                                                             Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream0-00200.m4s'                                                                                                                            Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00200.m4s'                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/manifest?X-Plex-Http-Pipeline=infinite' for writing                                      UnicornTranscoder:Transcoder  +3s                                                                                                                                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'chunk-stream0-00201.m4s.tmp' for writing                                                                                                                                                               UnicornTranscoder:Transcoder  +0ms                                                                                                                                                                                                                                             UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'chunk-stream1-00201.m4s.tmp' for writing                                                                                                                                                               UnicornTranscoder:Transcoder  +34ms                                                                                                                                                                                                                                            UnicornTranscoder:Dash Serving chunk-stream0-00200.m4s for session o5nrmi7g0aii6va25wr0d865 +1s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream1-00200.m4s for session o5nrmi7g0aii6va25wr0d865 +1s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream1-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                              Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00201.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream1-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                              Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00201.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream0-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +125ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream0-00201.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream0-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +883ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream0-00201.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream1-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +663ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00201.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream0-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +1s                                                                                                                                                                              Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream0-00201.m4s'                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                  UnicornTranscoder:Transcoder  +9s                                                                                                                                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [segment @ 0xf038c0] Opening 'sub-chunk-00016' for writing                                                                                                                                                                        UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                                  UnicornTranscoder:Transcoder  +0ms                                                                                                                                                                                                                                             UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/manifest?X-Plex-Http-Pipeline=infinite' for writing                                      UnicornTranscoder:Transcoder  +1s                                                                                                                                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'chunk-stream0-00202.m4s.tmp' for writing                                                                                                                                                               UnicornTranscoder:Transcoder  +2ms                                                                                                                                                                                                                                             UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'chunk-stream1-00202.m4s.tmp' for writing                                                                                                                                                               UnicornTranscoder:Transcoder  +19ms                                                                                                                                                                                                                                            UnicornTranscoder:Transcoder FFMPEG(stderr): [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                  UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'sub-chunk-00017' for writing                                                                                                                                                                                        UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                                  UnicornTranscoder:Transcoder  +120ms                                                                                                                                                                                                                                           UnicornTranscoder:Dash Serving chunk-stream1-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream1-00202.m4s for session o5nrmi7g0aii6va25wr0d865 +267ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00202.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream1-00202.m4s for session o5nrmi7g0aii6va25wr0d865 +1s                                                                                                                                                                              Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00202.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream0-00201.m4s for session o5nrmi7g0aii6va25wr0d865 +1s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream1-00202.m4s for session o5nrmi7g0aii6va25wr0d865 +984ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00202.m4s'                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/manifest?X-Plex-Http-Pipeline=infinite' for writing                                      UnicornTranscoder:Transcoder  +4s                                                                                                                                                                                                                                              UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'chunk-stream0-00203.m4s.tmp' for writing                                                                                                                                                               UnicornTranscoder:Transcoder  +1ms                                                                                                                                                                                                                                             UnicornTranscoder:Transcoder FFMPEG(stderr): [dash @ 0xf00580] Opening 'chunk-stream1-00203.m4s.tmp' for writing                                                                                                                                                               UnicornTranscoder:Transcoder  +14ms                                                                                                                                                                                                                                            UnicornTranscoder:Transcoder FFMPEG(stderr): [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                  UnicornTranscoder:Transcoder  +935ms                                                                                                                                                                                                                                           UnicornTranscoder:Transcoder FFMPEG(stderr): [segment @ 0xf038c0] Opening 'sub-chunk-00018' for writing                                                                                                                                                                        UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                                  UnicornTranscoder:Transcoder  +0ms                                                                                                                                                                                                                                             UnicornTranscoder:Dash Serving chunk-stream1-00202.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                                UnicornTranscoder:Dash Serving chunk-stream0-00202.m4s for session o5nrmi7g0aii6va25wr0d865 +2s                                                                                                                                                                                UnicornTranscoder:Transcoder FFMPEG(stderr):     Last message repeated 1 times                                                                                                                                                                                                 UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'sub-chunk-00019' for writing                                                                                                                                                                                        UnicornTranscoder:Transcoder [segment @ 0xf038c0] Opening 'http://127.0.0.1:3000/video/:/transcode/session/o5nrmi7g0aii6va25wr0d865/ce71de4c-4c4b-4062-9e85-c739c6855215/seglist' for writing                                                                                  UnicornTranscoder:Transcoder  +3s                                                                                                                                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream1-00203.m4s for session o5nrmi7g0aii6va25wr0d865 +877ms                                                                                                                                                                           Error: ENOENT: no such file or directory, stat '/home/magiclegend/unicorn/UnicornTranscoder/cache/o5nrmi7g0aii6va25wr0d865/chunk-stream1-00203.m4s'                                                                                                                              UnicornTranscoder:Dash Serving chunk-stream1-00203.m4s for session o5nrmi7g0aii6va25wr0d865
```

**Comment by Maxou44** [comment]
FFmpeg is throttled, it's the normal behavior 😉

**Comment by MagicLegend** [comment]
Hmm, but that makes it too slow to transcode this worst-case scenario :') (2160p 10.7 Mb/s source) Can I un-throttle it?

**Comment by Maxou44** [comment]
Plex and 4k it's... complicated, we can only handle 2 uhd stream using 30€/mo server. Plex send the value about throttling itself, transcoder forwards the value. Maybe VMs limit performance or your I/O are too limited to stream the file?

**Comment by MagicLegend** [comment]
> Plex and 4k it's... complicated

Here we go :D

> we can only handle 2 uhd stream using 30€/mo server

But where would that be bottlenecked then? Not in transcoder power I assume?

> Plex send the value about throttling itself, transcoder forwards the value

And I can't trick the transcoder into not caring about what Plex wants, and just take full resources? (When the transcoding is active, it can be assumed that the machine has nothing else to do)

> Maybe VMs limit performance

Of course. As I said, there is a little overhead added. However, I can see that FFMPEG is not taking all the resources the VM offers; which I want it to do. I will be the one who throttles it by not giving all the host resources to the VM; I don't like it when software thinks for me :)

> or your I/O are too limited to stream the file?

No, however, I must say that enabling `CUSTOM_DOWNLOAD_FORWARD` seemed to help a little (I guess it removed some of the overhead added by the looping around of the files). However, the reading of this movie should be a relatively sequential reading operation; which HDD's aren't too bad at. This specific one reaches +-120MB/s in CrystalDiskMark. More than enough to push out a 10.7Mb/s file I would say :)

**Comment by MagicLegend** [comment]
> After a little bit of testing I've found the culprit: Tampermonkey. I was afraid of that... Could you do me a favour and try to install the extension in a Chrome 80 and test if the `start.mpd` request to the **transcoder** gets through?

Could you give this a try for me? Then I know if Tampermonkey in general is just incompatible, and I should ask the developer of Tampermonkey if he has an idea what's up :)

**Comment by Maxou44** [comment]
I have the same issue on my side, it's a tampermonkey related bug, it's like if the 307 request was blocked

**Comment by MagicLegend** [comment]
> > Plex send the value about throttling itself, transcoder forwards the value
> 
> And I can't trick the transcoder into not caring about what Plex wants, and just take full resources? (When the transcoding is active, it can be assumed that the machine has nothing else to do)
> 
> > Maybe VMs limit performance
> 
> Of course. As I said, there is a little overhead added. However, I can see that FFMPEG is not taking all the resources the VM offers; which I want it to do. I will be the one who throttles it by not giving all the host resources to the VM; I don't like it when software thinks for me :)
> 
> > or your I/O are too limited to stream the file?
> 
> No, however, I must say that enabling `CUSTOM_DOWNLOAD_FORWARD` seemed to help a little (I guess it removed some of the overhead added by the looping around of the files). However, the reading of this movie should be a relatively sequential reading operation; which HDD's aren't too bad at. This specific one reaches +-120MB/s sequential read in CrystalDiskMark. More than enough to push out a 10.7Mb/s file I would say :)

Hi Max,

Sorry for getting back at you this late. I see you closed the issue, which is okay. However, I stil have one question left: How can I un-throttle FFMPEG? Is this a stetting buried in Plex somewhere?

**Comment by Maxou44** [comment]
Why do you want un-throttle FFmpeg ? This limit is here to limit CPU usage and it will transcode at least in x1.0 😉
If you really want to do it, you need to patch ffmpeg arguments and remove -bufsize value if I'm right

**Comment by MagicLegend** [comment]
Well, because I couldn't view the aforementioned 10.7Mb/s x265 file, because the transcoding was too slow :-) For lower bitrates it seems fine (I've tested 2.5Mb/s x265). If there is no easy way to unlimit FFMPEG in doing it's thing, I guess I just won't be watching 4k UHD remotely ;-)

**Comment by Maxou44** [comment]
In this case, If you can stream classic 720p/1080p x264 files, your CPU is too slow or the speed to serve files over network is too slow to direct play the file

**Comment by MagicLegend** [comment]
CPU of the transcoder, or the CPU of plex? The first is not bottlenecked for sure, and I'm also sure that the HDD that the file is being read off is fast enough (otherwise I wouldn't be able to watch the movie normally). The only thing I can think of is that FFMPEG is not spreading the tasks across the different cores; as I can see it's only trying to use 3 at most out of the 6 it has available. It is also possible that the plex CPU is bottlenecked (raspberry pi 3b+); I should test that tonight. Network is with a 100/100 fiber connection not an issue for sure.

Also please note: I can **stream** the 4k UHD file perfectly fine (using the Plex app over the internal network it doesn't try to transcode), but when it has to be **transcoded** it bottlenecks somewhere. I can imagine that the transcoding adds overhead for the plex server; hence why I don't assume the plex CPU being able to handle it.

**Comment by Maxou44** [comment]
What is your transcoder hardware?

"but when it has to be transcoded it bottlenecks somewhere": You transcode inside the network or ouside it? If it's ouside maybe you don't have a 100M for upload)

**Comment by MagicLegend** [comment]
> What is your transcoder hardware?

i7-7700K (OC @ 4.9GHz)
32GB of Corsair Vengeance LPX RAM
Bootdrive & VM location is a Samsung 850 EVO
This specific content is located on a Seagate 4TB disk (ST4000M004); other content is on a WD Blue (WD40EZRZ).
VM Virtualization by VMWare Workstation Pro 15.5.2 build-15785246 (which is capable of using the full CPU; I know this for a fact from other Ubuntu-like VMs)

VM has access to 6 cores and 4GB of RAM.

> You transcode inside the network or ouside it? If it's ouside maybe you don't have a 100M for upload). Can we have speedtests ?

As I stated, 100/100; so 100mbit upload as well. Speedtest (fast.com):
![image](https://user-images.githubusercontent.com/3169104/80915992-1f4c0100-8d56-11ea-813a-f3f429b33a24.png)
(As seen from the host)
```
magiclegend@debian:~$ curl -s https://raw.githubusercontent.com/sivel/speedtest-cli/master/speedtest.py | python -
Retrieving speedtest.net configuration...
Testing from ISP (1.2.3.4)...
Retrieving speedtest.net server list...
Selecting best server based on ping...
Hosted by (...) [9.32 km]: 34.976 ms
Testing download speed................................................................................
Download: 84.18 Mbit/s
Testing upload speed................................................................................................
Upload: 115.77 Mbit/s
```
As seen from the transcoder VM, on the host of the above results.

```
pi@raspberrypi ~/Documents
% curl -s https://raw.githubusercontent.com/sivel/speedtest-cli/master/speedtest.py | python -
Retrieving speedtest.net configuration...
Testing from ISP (1.2.3.4)...
Retrieving speedtest.net server list...
Selecting best server based on ping...
Hosted by (...) [9.32 km]: 12.565 ms
Testing download speed................................................................................
Download: 90.30 Mbit/s
Testing upload speed................................................................................................
Upload: 94.52 Mbit/s
```
As seen from the Pi (Plex host, while Plex is running, not streaming)

> Can you draw us your architecture setup?

But of course. I hope this is enough detail:

![image](https://user-images.githubusercontent.com/3169104/80916589-eada4400-8d59-11ea-8523-87cda4a6c052.png)

The internal network is 100% 1 gigabit speed. Both the switch and router are capable of this (we will be upgrading to a higher plan in a month; so the router must be capable of at least gigabit speeds as well).

Right now I use the `CUSTOM_DOWNLOAD_FORWARD` flag on the LB; and I have mounted the media folders to the Transcoder VM. The transcoder indeed seems to access the media files directly from the mount (as the mount-paths show up in the logs).
